### PR TITLE
Add fuzz tests missing vendor changes

### PR DESCRIPTION
**Pull request by pooneh-m**
vendor_fixes is missing fuzz testing dependencies.
https://github.com/googleforgames/agones/issues/1098

### CODE REVIEW

**Comment by pooneh-m** [comment]
I am not how it was working before, but I assume there were local changes that never got pushed? Or I am missing something.

**Comment by roberthbailey** [comment]
@aLekSer - is there a reason that these files weren't included in https://github.com/googleforgames/agones/pull/1269?

**Comment by aLekSer** [comment]
@roberthbailey  We have all this files in a `vendor` directory, so it seems that we don't need to add duplicate files into `vendor_fixes`:
```
 git ls-files ../vendor/k8s.io/apimachinery/pkg/api/validation/
../vendor/k8s.io/apimachinery/pkg/api/validation/doc.go
../vendor/k8s.io/apimachinery/pkg/api/validation/generic.go
../vendor/k8s.io/apimachinery/pkg/api/validation/objectmeta.go
../vendor/k8s.io/apimachinery/pkg/api/validation/objectmeta_test.go
../vendor/k8s.io/apimachinery/pkg/api/validation/path/name.go
../vendor/k8s.io/apimachinery/pkg/api/validation/path/name_test.go
```
+ 5 files here:
```
git ls-files ../vendor/k8s.io/apimachinery/pkg/api/apitesting/ | wc -l
```

**Comment by aLekSer** [comment]
Found a diff in roundtrip.go file, need to check the version:
```
diff ~/go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/api/apitesting/roundtrip/roundtrip.go  ~/go/src/agones.dev/agones/vendor_fixes/k8s.io/apimachinery/pkg/api/apitesting//roundtrip/roundtrip.go
143,149c143,145
<
<               // FIXME: this is explicitly testing w/o protobuf which was failing if enabled
<               // the reason for that is that protobuf is not setting Kind and APIVersion fields
<               // during obj2 decode, the same then applies to DecodeInto obj3. My guess is we
<               // should be setting these two fields accordingly when protobuf is passed as codec
<               // to roundTrip method.
<               roundTripSpecificKind(t, gvk, scheme, codecFactory, fuzzer, nonRoundTrippableTypes, true)                                                                                                
---
>               t.Run(gvk.Group+"."+gvk.Version+"."+gvk.Kind, func(t *testing.T) {
>                       roundTripSpecificKind(t, gvk, scheme, codecFactory, fuzzer, nonRoundTrippableTypes, false)                                                                                       
>               })
166d161
<       t.Logf("round tripping %v", gvk)
234c229
<                       s := protobuf.NewSerializer(scheme, scheme, "application/arbitrary.content.type")                                                                                                
---
>                       s := protobuf.NewSerializer(scheme, scheme)
253,255d247
<       externalGoType := reflect.TypeOf(object).PkgPath()
<       t.Logf("\tround tripping external type %v %v", externalGVK, externalGoType)
<
263c255
<               roundTrip(t, scheme, protobuf.NewSerializer(scheme, scheme, "application/protobuf"), object)                                                                                             
---
>               roundTrip(t, scheme, protobuf.NewSerializer(scheme, scheme), object)
```

**Comment by pooneh-m** [comment]
If I understand correctly vendor_fixes/k8s.io/apimachinery replaces vendor/k8s.io/apimachinery: https://github.com/googleforgames/agones/blob/master/go.mod#L61

So we should have a copy of apitesting in vendor_fixes, correct?

**Comment by aLekSer** [comment]
First of all, I think we should approve this PR with the requirement to copy those files to `vendor` as wel (using `go mod vendor` command, which resets the main module's vendor directory). You are correct, and I have broken the `vendor_fixes` as it can be seen from this output.
```
go mod vendor -v
~/go/src/agones.dev/agones/test/sdk/nodejs/node_modules/agones                                                                                         
go: finding k8s.io/apimachinery v0.17.2
agones.dev/agones/pkg/apis/agones/v1 imports
        k8s.io/apimachinery/pkg/api/validation: module k8s.io/apimachinery@latest (v0.17.2) found, but does not contain package k8s.io/apimachinery/pkg/api/validation                                   
agones.dev/agones/pkg/apis/agones/v1 imports
        k8s.io/apimachinery/pkg/apis/meta/v1/validation: module k8s.io/apimachinery@latest (v0.17.2) found, but does not contain package k8s.io/apimachinery/pkg/apis/meta/v1/validation
```
Second is that statement taken from [Go cmd tips](https://tip.golang.org/cmd/go/#hdr-Modules_and_vendoring). In which we skip using 
```
To build using the main module's top-level vendor directory to satisfy dependencies (disabling use of the usual network sources and local caches), use 'go build -mod=vendor'. Note that only the main module's top-level vendor directory is used; vendor directories in other locations are still ignored.
```

File: vendor_fixes/k8s.io/apimachinery/pkg/api/apitesting/roundtrip/roundtrip.go [hunk 1]
```diff
@@ -0,0 +1,415 @@
+/*
```
Comment (aLekSer) [inline vendor_fixes/k8s.io/apimachinery/pkg/api/apitesting/roundtrip/roundtrip.go]: Please add this and other files to a `vendor` directory as well. This file has diff with what I have added to `vendor`. Please use the version which is taken from k8s.io/apimachinery v0.0.0-20191004074956-01f8b7d1121a

Comment (pooneh-m) [reply vendor_fixes/k8s.io/apimachinery/pkg/api/apitesting/roundtrip/roundtrip.go]: I am not sure if I understand this comment. Are you saying apimachinary copy in vendor is not the same as the version specified in mod?

When I run `go mod vendor` roundtrip.go doesn't change. However, there are other changes:
```
	deleted:    vendor/github.com/google/btree/btree_mem.go
	deleted:    vendor/golang.org/x/sys/unix/mkasm_darwin.go
	deleted:    vendor/golang.org/x/sys/unix/mkpost.go
	deleted:    vendor/golang.org/x/sys/unix/mksyscall.go
	deleted:    vendor/golang.org/x/sys/unix/mksyscall_aix_ppc.go
	deleted:    vendor/golang.org/x/sys/unix/mksyscall_aix_ppc64.go
	deleted:    vendor/golang.org/x/sys/unix/mksyscall_solaris.go
	deleted:    vendor/golang.org/x/sys/unix/mksysnum.go
	deleted:    vendor/golang.org/x/sys/unix/types_aix.go
	deleted:    vendor/golang.org/x/sys/unix/types_darwin.go
	deleted:    vendor/golang.org/x/sys/unix/types_dragonfly.go
	deleted:    vendor/golang.org/x/sys/unix/types_freebsd.go
	deleted:    vendor/golang.org/x/sys/unix/types_netbsd.go
	deleted:    vendor/golang.org/x/sys/unix/types_openbsd.go
	deleted:    vendor/golang.org/x/sys/unix/types_solaris.go
	deleted:    vendor/golang.org/x/text/unicode/bidi/gen.go
	deleted:    vendor/golang.org/x/text/unicode/bidi/gen_ranges.go
	deleted:    vendor/golang.org/x/text/unicode/bidi/gen_trieval.go
	deleted:    vendor/golang.org/x/text/unicode/norm/maketables.go
	deleted:    vendor/golang.org/x/text/unicode/norm/triegen.go
	deleted:    vendor/golang.org/x/text/width/gen.go
	deleted:    vendor/golang.org/x/text/width/gen_common.go
	deleted:    vendor/golang.org/x/text/width/gen_trieval.go
	deleted:    vendor/golang.org/x/tools/go/gcexportdata/main.go
	deleted:    vendor/golang.org/x/tools/imports/mkindex.go
	deleted:    vendor/golang.org/x/tools/imports/mkstdlib.go
	deleted:    vendor/k8s.io/apimachinery/pkg/api/apitesting/fuzzer/valuefuzz_test.go
	deleted:    vendor/k8s.io/apimachinery/pkg/api/validation/objectmeta_test.go
	deleted:    vendor/k8s.io/apimachinery/pkg/api/validation/path/name.go
	deleted:    vendor/k8s.io/apimachinery/pkg/api/validation/path/name_test.go
	deleted:    vendor/k8s.io/apimachinery/pkg/apis/meta/v1/validation/validation_test.go
```

**Comment by aLekSer** [comment]
I mean difference between vendor and vendor_fixes for apimachinery they should be equal. Thought `go mod vendor` would create missing files and also fixed that difference.

**Comment by pooneh-m** [comment]
> I mean difference between vendor and vendor_fixes for apimachinery they should be equal. Thought `go mod vendor` would create missing files and also fixed that difference.

Exactly, with this PR vendor_fix has the copy of required files for round trip fuzz testing from vendor. The vendor is based on `k8s.io/apimachinery v0.0.0-20191004074956-01f8b7d1121a`

**Comment by aLekSer** [comment]
One more thing I have found is:
```
go mod tidy                                                                                      
agones.dev/agones/cmd/alloc imports
        k8s.io/client-go/rest tested by
        k8s.io/client-go/rest.test imports
        k8s.io/apimachinery/pkg/util/httpstream: module k8s.io/apimachinery@latest (v0.17.2) found, but does not contain package k8s.io/apimachinery/pkg/util/httpstream    
```

**Comment by pooneh-m** [comment]
> One more thing I have found is:
> 
> ```
> go mod tidy                                                                                      
> agones.dev/agones/cmd/allocator imports
>         k8s.io/client-go/rest tested by
>         k8s.io/client-go/rest.test imports
>         k8s.io/apimachinery/pkg/util/httpstream: module k8s.io/apimachinery@latest (v0.17.2) found, but does not contain package k8s.io/apimachinery/pkg/util/httpstream    
> ```
> 
> With no directory in the path `./vendor_fixes/k8s.io/apimachinery/pkg/util/httpstream`
> I think we could add them as part of this PR to be consistent.

Good find. There is no k8s.io/apimachinery/pkg/util/httpstream to copy over. We need to upgrade apimachinery. I created this issue https://github.com/googleforgames/agones/issues/1309 to track that.

**Pull request merged**